py/src/braintrust/logger.py

```diff
@@ -1649,6 +1722,46 @@ def _validate_and_sanitize_experiment_log_full_args(event: Mapping[str, Any], ha
     return event
 
 
+def _deep_copy_event(event: Mapping[str, Any]) -> Dict[str, Any]:
+    """
+    Creates a deep copy of the given event. Replaces references to user objects
+    with placeholder strings to ensure serializability, except for `Attachment`
+    objects, which are preserved and not deep-copied.
+    """
+
+    def _deep_copy_object(v: Any) -> Any:
+        if isinstance(v, Span):
+            return "<span>"
+        elif isinstance(v, Experiment):
+            return "<experiment>"
+        elif isinstance(v, Dataset):
+            return "<dataset>"
+        elif isinstance(v, Logger):
+            return "<logger>"
+        elif isinstance(v, Attachment):
+            return v
+        else:
+            # No need to handle primitives explicitly because deepcopy will do
+            # it for us.
+            return copy.deepcopy(v)
+
+    ret: Dict[str, Any] = {}
+
+    for k, v in event.items():
+        k = str(k)
+
+        if isinstance(v, Mapping):
+            v = _deep_copy_event(v)
+        elif isinstance(v, (List, Tuple, Set)):
```

Comment:
Maybe a tiny bit more pythonic to use `collections.abc.Sequence` instead of `List/Tuple`.

Reply:
`Sequence` also includes `str`, `bytes`, `bytearray`, `array`, etc. The condition would have to be something like:

```py
isinstance(v, (Sequence, Set)) and not isinstance(v, (str, Buffer))
```

another option:
```py
isinstance(v, Iterable) and not isinstance(v, (str, Buffer, Mapping))
```

At that point, I thought it would be less error-prone to have a best-effort list of allowed types rather than figure out the comprehensive list of disallowed types.